server/src/main/java/org/elasticsearch/search/aggregations/TopBucketBuilder.java

```diff
@@ -20,26 +21,47 @@
 /**
  * Merges many buckets into the "top" buckets as sorted by {@link BucketOrder}.
  */
-public class TopBucketBuilder<B extends InternalMultiBucketAggregation.InternalBucket> {
-    private final Consumer<DelayedBucket<B>> nonCompetitive;
-    private final PriorityQueue<DelayedBucket<B>> queue;
+public abstract class TopBucketBuilder<B extends InternalMultiBucketAggregation.InternalBucket> {
+    /**
+     * The number of buckets required before we switch to the
+     * {@link BufferingTopBucketBuilder}. If we need fewer buckets we use
+     * {@link PriorityQueueTopBucketBuilder}.
+     */
+    static final int USE_BUFFERING_BUILDER = 1024;
```

Comment:
I am really curious how you came up with this number.

Reply:
I looked into my heart and thought "32k is fine to over allocate".  And I thought "this is much bigger than the default of 10 and in the ball park of "too many bars to have in a bar chart". But, fundamentally, it was a guess because I didn't think it was worth spending a lot of time trying to come up with the perfect number. I will add a comment about that.